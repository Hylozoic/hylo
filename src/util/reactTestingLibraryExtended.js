import React from 'react'
import { MemoryRouter } from 'react-router'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import rootReducer from 'store/reducers'
import createMiddleware from 'store/middleware'

export function generateStore (history, initialState) {
  return createStore(rootReducer, initialState, createMiddleware(history))
}

export const AllTheProviders = (store) => ({ children }) => {
  return (
    <Provider store={store || createStore(() => {})}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
}

// re-export everything

const rtl = require('react-testing-library')
const customRender = (ui, options, providersFunc) =>
  rtl.render(ui, { wrapper: providersFunc, ...options })

module.exports = {
  ...rtl,
  render: customRender
}
