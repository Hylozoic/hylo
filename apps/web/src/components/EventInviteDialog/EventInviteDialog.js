import React, { useState, useEffect } from 'react'
import { useInView } from 'react-cool-inview'
import { useTranslation } from 'react-i18next'
import * as Dialog from '@radix-ui/react-dialog'
import { humanResponse } from '@hylo/presenters/EventInvitationPresenter'
import Button from 'components/Button'
import CheckBox from 'components/CheckBox'
import Loading from 'components/Loading'
import TextInput from 'components/TextInput'
import { bgImageStyle, cn } from 'util/index'
import { CENTER_COLUMN_ID } from 'util/scrolling'

import styles from './EventInviteDialog.module.scss'

const pageSize = 30

const EventInviteDialog = ({
  eventTitle,
  fetchPeople,
  forGroups,
  eventInvitations,
  people,
  eventId,
  onClose,
  invitePeopleToEvent,
  pending
}) => {
  const [invitedIds, setInvitedIds] = useState([])
  const [searchTerm, setSearchTerm] = useState('')
  const [pageFetched, setPageFetched] = useState(0)

  const toggleInvite = id => (invitedIds.includes(id))
    ? setInvitedIds(invitedIds.filter(invitedId => invitedId !== id))
    : setInvitedIds(invitedIds.concat([id]))

  const onSearchChange = ({ target: { value } }) => setSearchTerm(value)
  const { t } = useTranslation()

  useEffect(() => {
    const fetch = () => {
      const forGroupIds = forGroups.map(c => c.id).filter(c => c !== 'public')
      fetchPeople({ autocomplete: searchTerm, groupIds: forGroupIds, first: pageSize, offset: 0 })
      setPageFetched(pageSize)
    }
    fetch()
  }, [searchTerm])

  const { observe } = useInView({
    onEnter: () => {
      const fetch = () => {
        const forGroupIds = forGroups.map(c => c.id).filter(c => c !== 'public')
        fetchPeople({ autocomplete: searchTerm, groupIds: forGroupIds, first: pageSize, offset: pageFetched })
        setPageFetched(pageFetched + pageSize)
      }
      fetch()
    }
  })

  const getFilteredInviteSuggestions = () => {
    return people.filter(person => {
      return !eventInvitations.map(ei => ei.id).includes(person.id) &&
      person.name.toLowerCase().includes(searchTerm.toLowerCase())
    })
  }

  const submit = () => {
    invitePeopleToEvent(eventId, invitedIds)
    onClose()
  }

  const filteredInviteSuggestions = getFilteredInviteSuggestions()

  const inviteButtonLabel = invitedIds.length === 0
    ? t('Select people to invite')
    : invitedIds.length === 1
      ? t('Invite 1 person')
      : t('Invite {{invitedIds.length}} people', { invitedIds })

  return (
    <Dialog.Root defaultOpen onOpenChange={onClose}>
      <Dialog.Portal container={document.getElementById(CENTER_COLUMN_ID)}>
        <Dialog.Overlay className='EventInviteDialog-Overlay bg-black/50 absolute left-0 right-0 bottom-0 grid place-items-center overflow-y-auto z-[100] h-full backdrop-blur-sm p-2'>
          <Dialog.Content className='EventInviteDialog-Content min-w-[300px] w-full bg-background p-3 rounded-md z-[41] max-w-[750px] outline-none relative'>
            <Dialog.Title className='text-xl font-semibold leading-none tracking-tight'>Invite to event: {eventTitle}</Dialog.Title>
            <Dialog.Description className='sr-only'>Invite group members to event: {eventTitle}</Dialog.Description>
            <div className={styles.container}>
              <Search onChange={onSearchChange} />
              <div className={styles.inviteSuggestions}>
                {filteredInviteSuggestions.map((invitee, idx) => (
                  <InviteeRow
                    key={invitee.id}
                    person={invitee}
                    ref={idx === filteredInviteSuggestions.length - 10 ? observe : null}
                    selected={invitedIds.includes(invitee.id)}
                    onClick={() => toggleInvite(invitee.id)}
                  />))}
                <div className={cn(styles.row)}>
                  <div className={cn(styles.col)} style={{ height: '40px' }}>
                    {pending && <div><Loading /></div>}
                  </div>
                </div>
              </div>

              <div className='font-bold pb-2 mt-6'>{t('Already Invited')}</div>
              <div className={styles.alreadyInvited}>
                {eventInvitations.map(eventInvitation =>
                  <InviteeRow
                    person={eventInvitation}
                    showResponse
                    key={eventInvitation.id}
                  />)}
              </div>
              <Button
                small
                className={styles.inviteButton}
                label={inviteButtonLabel}
                onClick={submit}
                disabled={invitedIds.length === 0}
              />
            </div>
          </Dialog.Content>
        </Dialog.Overlay>
      </Dialog.Portal>
    </Dialog.Root>
  )
}

export const InviteeRow = React.forwardRef((props, ref) => {
  const { t } = useTranslation()
  const { person, selected, showResponse, onClick } = props
  const { name, avatarUrl, response } = person
  return (
    <div ref={ref} className={cn(styles.row)} onClick={onClick}>
      <div className={styles.col}>
        <div className={styles.avatar} style={bgImageStyle(avatarUrl)} />
      </div>
      <div className={styles.col}>
        {name}
      </div>
      {!showResponse && (
        <div className={cn(styles.col, styles.check)}>
          <CheckBox checked={selected} noInput />
        </div>
      )}
      {showResponse && response && (
        <div className={cn(styles.col, styles.response)}>
          {t(humanResponse(response))}
        </div>
      )}
    </div>
  )
})

export function Search ({ onChange }) {
  const { t } = useTranslation()
  return (
    <div className={styles.search}>
      <TextInput
        theme={styles}
        inputRef={x => x && x.focus()}
        placeholder={t('Search members')}
        onChange={onChange}
      />
    </div>
  )
}

export default EventInviteDialog
