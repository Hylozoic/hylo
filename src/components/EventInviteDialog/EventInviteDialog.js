import React, {useState, useEffect} from 'react'
import cx from 'classnames'
import { bgImageStyle } from 'util/index'
import ModalDialog from 'components/ModalDialog'
import CheckBox from 'components/CheckBox'
import Button from 'components/Button'
import styles from './EventInviteDialog.scss'
import { humanResponse } from 'store/models/EventInvitation'
import TextInput from 'components/TextInput'

export default function EventInviteDialog ({
  fetchPeople,
  forGroups,
  eventInvitations,
  people,
  eventId,
  onClose,
  invitePeopleToEvent
}) {
  const [invitedIds, setInvitedIds] = useState([])
  const [searchTerm, setSearchTerm] = useState('')

  const toggleInvite = id => (invitedIds.includes(id)) 
    ? setInvitedIds(invitedIds.filter(invitedId => invitedId !== id))
    : setInvitedIds(invitedIds.concat([id])) 

  const onSearchChange = ({ target: { value } }) => setSearchTerm(value)

  useEffect(() => {
      const forGroupIds = forGroups.map(c => c.id)
      fetchPeople(searchTerm, forGroupIds)
  }, [searchTerm])

  const getFilteredInviteSuggestions = () => {
    return people.filter(person => {
      return !eventInvitations.map(ei => ei.id).includes(person.id) &&
      person.name.toLowerCase().includes(searchTerm.toLowerCase())
    })
  }

  const submit = () => {
    invitePeopleToEvent(eventId, invitedIds)
    onClose()
  }

  const filteredInviteSuggestions = getFilteredInviteSuggestions()

  const inviteButtonLabel = invitedIds.length === 0
    ? 'Select people to invite'
    : invitedIds.length === 1
      ? 'Invite 1 person'
      : `Invite ${invitedIds.length} people`

  return <ModalDialog key='event-invite-dialog'
    closeModal={onClose}
    modalTitle={`Invite`}
    showCancelButton={false}
    showSubmitButton={false}
    style={{ width: '100%', maxWidth: '620px' }}>
    <div styleName='container'>
      <Search onChange={onSearchChange} />
      <div styleName='inviteSuggestions'>
        {filteredInviteSuggestions.map(invitee => <InviteeRow
          key={invitee.id}
          person={invitee}
          selected={invitedIds.includes(invitee.id)}
          onClick={() => toggleInvite(invitee.id)}
        />)}
      </div>
      <div styleName='alreadyInvitedLabel'>Already Invited</div>
      <div styleName='alreadyInvited'>
        {eventInvitations.map(eventInvitation =>
          <InviteeRow
            person={eventInvitation}
            showResponse
            key={eventInvitation.id}
          />)}
      </div>
      <Button
        small
        styleName='inviteButton'
        label={inviteButtonLabel}
        onClick={submit}
        disabled={invitedIds.length === 0} />
    </div>
  </ModalDialog>
}

export function InviteeRow ({ person, selected, showResponse, onClick }) {
  const { name, avatarUrl, response } = person

  return <div styleName={cx('row')} onClick={onClick}>
    <div styleName='col'>
      <div styleName='avatar' style={bgImageStyle(avatarUrl)} />
    </div>
    <div styleName='col'>
      {name}
    </div>
    {!showResponse && <div styleName='col check'>
      <CheckBox checked={selected} noInput />
    </div>}
    {showResponse && response && <div styleName='col response'>
      {humanResponse(response)}
    </div>}
  </div>
}

export function Search ({ onChange }) {
  return <div styleName='search'>
    <TextInput theme={styles}
      inputRef={x => x && x.focus()}
      placeholder='Search members'
      onChange={onChange} />
  </div>
}
