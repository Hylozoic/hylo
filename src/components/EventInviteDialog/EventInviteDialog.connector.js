import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { debounce } from 'lodash/fp'
import fetchPeople from 'store/actions/fetchPeople'
import { invitePeopleToEvent, peopleSelector } from './EventInviteDialog.store'

export function mapStateToProps (state, props) {
  const people = peopleSelector(state, props)

  return {
    people
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    ...bindActionCreators({
      invitePeopleToEvent
    }, dispatch),
    fetchPeople: debounce(300, (args) => dispatch(fetchPeople(args)))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
