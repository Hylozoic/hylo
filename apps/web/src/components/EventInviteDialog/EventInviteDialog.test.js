import React from 'react'
import { render, screen, fireEvent } from 'util/testing/reactTestingLibraryExtended'
import EventInviteDialog, { InviteeRow, Search } from './EventInviteDialog'

describe('EventInviteDialog', () => {
  it('renders correctly', () => {
    const props = {
      onClose: jest.fn(),
      eventInvitations: [{ id: 1 }, { id: 2 }, { id: 3 }],
      people: [{ id: 1, name: 'Alice' }, { id: 4, name: 'Bob' }, { id: 5, name: 'Charlie' }],
      forGroups: [],
      fetchPeople: jest.fn(),
      eventId: '123',
      invitePeopleToEvent: jest.fn(),
      pending: false
    }
    render(<EventInviteDialog {...props} />)

    expect(screen.getByText('Invite')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('Search members')).toBeInTheDocument()
    expect(screen.getByText('Already Invited')).toBeInTheDocument()
    expect(screen.getByText('Select people to invite')).toBeInTheDocument()
  })
})

describe('InviteeRow', () => {
  it('renders correctly', () => {
    const props = {
      person: { id: 1, name: 'John', avatarUrl: 'john.png', response: 'not needed' },
      selected: true,
      onClick: jest.fn()
    }
    render(<InviteeRow {...props} />)

    expect(screen.getByText('John')).toBeInTheDocument()
    expect(screen.getByRole('checkbox')).toBeChecked()
  })

  it('renders correctly with showResponse', () => {
    const props = {
      person: { id: 1, name: 'Jane', avatarUrl: 'jane.png', response: 'needed!' },
      selected: true,
      onClick: jest.fn(),
      showResponse: true
    }
    render(<InviteeRow {...props} />)

    expect(screen.getByText('Jane')).toBeInTheDocument()
    expect(screen.getByText('needed!')).toBeInTheDocument()
    expect(screen.queryByRole('checkbox')).not.toBeInTheDocument()
  })
})

describe('Search', () => {
  it('renders correctly and handles input', () => {
    const onChange = jest.fn()
    render(<Search onChange={onChange} />)

    const input = screen.getByPlaceholderText('Search members')
    expect(input).toBeInTheDocument()

    fireEvent.change(input, { target: { value: 'test' } })
    expect(onChange).toHaveBeenCalledTimes(1)
  })
})
