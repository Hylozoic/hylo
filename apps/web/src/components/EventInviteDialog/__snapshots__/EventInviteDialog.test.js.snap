// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InviteeRow renders correctly 1`] = `
<div
  data-stylename="row"
  onClick={[Function]}
>
  <div
    data-stylename="col"
  >
    <div
      data-stylename="avatar"
      style={
        Object {
          "backgroundImage": "url(j.png)",
        }
      }
    />
  </div>
  <div
    data-stylename="col"
  >
    j
  </div>
  <div
    data-stylename="col check"
  >
    <CheckBox
      checked={true}
      noInput={true}
    />
  </div>
</div>
`;
